src/shared/bpf-program.c

```diff
@@ -325,7 +326,18 @@ int bpf_map_new(
         for (size_t i = 0; i < sizeof(attr.map_name) - 1 && *n; i++, n++)
                 attr.map_name[i] = strchr(ALPHANUMERICAL ".", *n) ? *n : '_';
 
-        return RET_NERRNO(bpf(BPF_MAP_CREATE, &attr, sizeof(attr)));
+        fd = RET_NERRNO(bpf(BPF_MAP_CREATE, &attr, sizeof(attr)));
```

Comment:
this is not noing to work. this here is using the war kernel syscall, not the new libbpf stuff. THe kernel syscall got this right, the libbpf stuff didn't. Hence you are fixing this at the wrong place.

I don't think we can workaround this at all in systemd, because libbpf keeps the fd around internally in its objects.

Reply:
hmm, you could probably hack it by adding

```
fd_cloexec(sym_bpf_map__fd(…), true);
```

whenever we load a bpf program via libbpf, but that is really ugly, and doesn't scale...

Reply:
background: pid1 currently has 6 bpf programs:

1. bpf-restrict-fs (implemented via libbpf)
2. bpf-restrict-ifaces (ditto)
3. bpf-socket-bind (ditto)
4. bpf-devices (implemented via raw ebpf assembly and raw bpf syscalls)
5. bpf-firewall (ditto)
6. bpf-foreign (ditto)

We should of course port 4…6 to libbpf too, but so far noone wanted to spend the time on it.

hence, it's the first four that have the O_CLOEXEC issue.

Reply:
I think you're right. I booted with the patch and the issue still occurs.